{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "wanted-rebate",
   "metadata": {},
   "source": [
    "# 1. set up"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "greenhouse-rotation",
   "metadata": {},
   "source": [
    "## 1.1 library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "narrow-jersey",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "print(\"print version\")\n",
    "print(sys.version)\n",
    "\n",
    "import os\n",
    "import time\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# for plotting\n",
    "import matplotlib.patches as patches # for plotting figure\n",
    "from matplotlib.patches import Patch\n",
    "import gc # for collecting garbage\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib.font_manager as fm\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn import preprocessing\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import cvxpy as cp\n",
    "\n",
    "from helper_plot import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "inner-promise",
   "metadata": {},
   "source": [
    "## 1.2 data read"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "advisory-acoustic",
   "metadata": {},
   "source": [
    "### 1.2.1. read variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73d4376",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read df_meta_population\n",
    "path_population = '../data/population/'\n",
    "file_population = path_population + 'meta_sub.tsv'\n",
    "fn_file_exists(file_population)\n",
    "df_meta_population = pd.read_csv(file_population, sep = '\\t')\n",
    "\n",
    "vec_sub = ['GBR', 'FIN', 'CHS', 'PUR', 'CDX',\n",
    "           'CLM', 'IBS', 'PEL', 'PJL', 'KHV',\n",
    "           'ACB', 'GWD', 'ESN', 'BEB', 'MSL',\n",
    "           'STU', 'ITU', 'CEU', 'YRI', 'CHB',\n",
    "           'JPT', 'LWK', 'ASW', 'MXL', 'TSI',\n",
    "           'GIH']\n",
    "\n",
    "dict_sub_index = {}\n",
    "for i in range(26):\n",
    "    dict_sub_index[vec_sub[i]] = i\n",
    "    \n",
    "#df_meta_population['list_indices']\n",
    "dict_vec_GT = {\n",
    "    'GBR': 13, 'FIN': 65, 'CHS': 193, 'PUR': 315, 'CDX': 380,\n",
    "    'CLM': 452, 'IBS': 639, 'PEL': 679, 'PJL': 1219, 'KHV': 819,\n",
    "    'ACB': 859, 'GWD': 1248, 'ESN': 1463, 'BEB': 1769, 'MSL': 1526,\n",
    "    'STU': 1817, 'ITU': 1857, 'CEU': 2164, 'YRI': 2343, 'CHB': 2372,\n",
    "    'JPT': 2522, 'LWK': 2592, 'ASW': 2860, 'MXL': 2828, 'TSI': 2998,\n",
    "    'GIH': 3105\n",
    "}\n",
    "\n",
    "\n",
    "dict_super = {row['SUB']: row['POP'] for index, row in df_meta_population.iterrows()}\n",
    "\n",
    "vec_mixed = [\n",
    "    'ACB_BEB', 'CHS_PJL', 'CLM_YRI', 'FIN_CHS', 'GBR_GWD', 'IBS_YRI',\n",
    "    'PUR_CEU', 'CHS_KHV', 'CLM_CLM', 'ESN_MXL', 'GBR_GIH', 'IBS_ESN', \n",
    "    'PJL_ITU', 'YRI_YRI']\n",
    "\n",
    "vec_sup_full = vec_pop_ordered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e79686",
   "metadata": {},
   "source": [
    "# 2. bone marrow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515fb191",
   "metadata": {},
   "source": [
    "## 2.1. functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f84f4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to get the % of present genotypes\n",
    "def fn_get_prop(input_temp):\n",
    "    # Assuming dict_GT['210521_MNC_1'] is your NumPy array\n",
    "    array_values, counts = np.unique(input_temp, return_counts=True)\n",
    "\n",
    "    # Create a dictionary to store the value counts\n",
    "    value_counts_dict = dict(zip(array_values, counts))\n",
    "\n",
    "    # Now you can access the value counts of each element\n",
    "    return((value_counts_dict[0.0] + value_counts_dict[0.5] + value_counts_dict[1.0]) / n_mut)\n",
    "\n",
    "def plot_bar_bm(df_prob, label_sample = \"\", loc_legend = \"left\", \n",
    "                bool_save_plot = False, path_plot = \"\", name_plot = \"\", size_plot_x = 3.0,\n",
    "                size_plot_y = 2.0, pad_x_tick = 0.035, total_width = 1):\n",
    "    n_sample = df_prob.shape[1]\n",
    "    df_prob = df_prob.loc[vec_pop_ordered]\n",
    "    fig, ax = plt.subplots(figsize=(size_plot_x, size_plot_y))  # Create a figure and an axes.\n",
    "\n",
    "    # Set the background color\n",
    "    fig.set_facecolor('white')  # Set the figure background to white\n",
    "    ax.set_facecolor('white')   # Set the axes background to white\n",
    "    \n",
    "    # Bar plot settings\n",
    "    # total_width = 1  # Total width for a group of bars\n",
    "    bar_width = total_width / n_sample\n",
    "    \n",
    "    # bar_width = 1.0 / n_sample  # Width of individual bars\n",
    "    # total_width = n_sample * bar_width  # Total width for a group of bars\n",
    "    ind = np.arange(len(df_prob))\n",
    "    \n",
    "    # Plot each bar\n",
    "    for i, col in enumerate(df_prob.columns):\n",
    "        if (i > 0):\n",
    "            i_color = i + 2\n",
    "            if (i > 1):\n",
    "                i_color = i_color + 1\n",
    "        else:\n",
    "            i_color = i\n",
    "                \n",
    "        offset = (i + 1/2) * bar_width - total_width / 2  # Center bars around each tick\n",
    "        ax.bar(ind + offset, df_prob[col], \n",
    "               bar_width, # bar width # bar_width + n_sample/20\n",
    "               color = plt.cm.viridis(np.linspace(0, 1, n_sample + 3))[i_color], \n",
    "               edgecolor = plt.cm.viridis(np.linspace(0, 1, n_sample + 3))[i_color], \n",
    "               label = col, linewidth = 0)\n",
    "        \n",
    "    # Set x-ticks to the center of each group of bars\n",
    "    ax.set_xticks(ind)\n",
    "    ax.set_xticklabels(df_prob.index, rotation='vertical', fontsize=size_font)\n",
    "\n",
    "    plt.xlim(-0.5, 25.5)\n",
    "    plt.xlabel('Populations', fontsize = size_font)\n",
    "    plt.ylabel('Predicted Probabilities', fontsize = size_font)\n",
    "    plt.yticks(np.arange(0, 1.1, 0.2))  # Adjust the range and step as needed\n",
    "\n",
    "    plt.tick_params(axis = 'both', which = 'major', labelsize = size_font)\n",
    "\n",
    "    # gridlines in the y-axis\n",
    "    # ax.grid(True, which='both', axis='y', linestyle='--', linewidth=0.5, color = \"black\")\n",
    "\n",
    "    # get handles and labels from the old legend\n",
    "    old_handles, old_labels = ax.get_legend_handles_labels()\n",
    "\n",
    "    # create new handles for the new legend entries\n",
    "    new_handles = [patches.Rectangle((0, 0), 1, 1, color = dict_color_super[region_i]) for region_i in vec_sup]\n",
    "    new_labels = vec_sup_full\n",
    "\n",
    "    if label_sample == \"\":\n",
    "        label_sample = [\"sample\" + str(i) for i in range(1, n_sample + 1)]\n",
    "    \n",
    "    labels_donor = label_sample\n",
    "    # combine old and new handles and labels\n",
    "    handles = old_handles + new_handles\n",
    "    # labels = labels_donor + new_labels\n",
    "    labels = labels_donor[:n_sample]\n",
    "\n",
    "    # create the new legend\n",
    "    legend_bar = ax.legend(handles, labels, loc = 'upper ' + loc_legend, fontsize = size_font,\n",
    "                           frameon=True, facecolor='white', edgecolor = \"black\")\n",
    "    \n",
    "    # legend_bar.set_title(title_plot + \" (% of GT Reads)\", prop={\"size\": size_font}) \n",
    "    ax.add_artist(legend_bar)\n",
    "\n",
    "    for label in ax.get_xticklabels():  \n",
    "        pop_temp = label.get_text()\n",
    "        color_temp = dict_color_super[df_meta_unique[df_meta_unique['SUP'] == pop_temp]['POP'].values[0]]\n",
    "        label.set_bbox(dict(facecolor = color_temp, edgecolor = 'None', alpha = 0.5, pad = 0.5))  # change color as needed\n",
    "    \n",
    "    # Adjusting space between the x-axis and its tick labels\n",
    "    ## 1. Reduce pad to bring labels closer\n",
    "    ax.tick_params(axis = 'x', which = 'major', pad = 0.001)\n",
    "\n",
    "    ## 2. Manually adjust the position of x-axis tick labels\n",
    "    for label in ax.get_xticklabels():\n",
    "        label.set_y(label.get_position()[1] + pad_x_tick)\n",
    "        \n",
    "    # Reduce empty space inside the plot area\n",
    "    ax.margins(x = 0.001)  # Reduce the x margins (Adjust the value as needed)\n",
    "\n",
    "    ax.spines['top'].set_visible(True)\n",
    "    ax.spines['right'].set_visible(True)\n",
    "    ax.spines['bottom'].set_visible(True)\n",
    "    ax.spines['left'].set_visible(True)\n",
    "\n",
    "    ax.spines['top'].set_color('black')\n",
    "    ax.spines['right'].set_color('black')\n",
    "    ax.spines['bottom'].set_color('black')\n",
    "    ax.spines['left'].set_color('black')\n",
    "\n",
    "    ax.spines['top'].set_linewidth(1)\n",
    "    ax.spines['right'].set_linewidth(1)\n",
    "    ax.spines['bottom'].set_linewidth(1)\n",
    "    ax.spines['left'].set_linewidth(1)\n",
    "    \n",
    "    ax.yaxis.set_ticks([])\n",
    "    ax.yaxis.set_ticklabels([]) \n",
    "    \n",
    "    # Reduce padding and adjust layout\n",
    "    plt.tight_layout(pad=0.4, w_pad=0.3, h_pad=0.3)  # Adjust padding parameters as needed\n",
    "\n",
    "    # Adjust subplot parameters to reduce unnecessary margins\n",
    "    plt.subplots_adjust(left=0.05, right=0.975, top=0.975, bottom=0.10)\n",
    "    \n",
    "    if (bool_save_plot):\n",
    "        fn_path_exists(path_plot)\n",
    "        print(f'<plot_bar_admixed> saving at path_plot: {path_plot}')\n",
    "        plt.savefig(f'{fn_ensure_slash(path_plot)}{name_plot}_vertical.pdf', \n",
    "                    format = 'pdf', dpi = 1200, bbox_inches = 'tight')\n",
    "        plt.savefig(f'{fn_ensure_slash(path_plot)}{name_plot}_vertical.jpeg', \n",
    "                    format = 'jpeg', dpi = 1200, bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6d9c39c",
   "metadata": {},
   "source": [
    "## 2.2. data process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f97fdc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_p_site_present = {\n",
    "    'BM_1': 5.87, # white\n",
    "    'BM_2': 6.01, # black\n",
    "    'BM_3': 3.14, # asian\n",
    "    'BM_4': 2.57, # indian\n",
    "    'BM_5': 8.97, # asian\n",
    "    'BM_6': 13.0, # white\n",
    "    'BM_7': 7.43, # white\n",
    "    'BM_WGS_6': 99.3, # white\n",
    "    'BM_WGS_7': 97.6 # white\n",
    "}\n",
    "\n",
    "file_df_prob_input = '../data/results_BM/df_prob_BM_240429.tsv'\n",
    "fn_file_exists(file_df_prob_input)\n",
    "df_prob_input = pd.read_csv(\n",
    "    file_df_prob_input,\n",
    "    sep = '\\t', index_col = 0)\n",
    "\n",
    "df_meta = pd.DataFrame.from_dict(dict_p_site_present, orient='index', columns=['GT.read'])\n",
    "df_meta['ID_ind'] = list(df_meta.index)\n",
    "df_meta['origin'] = list(df_meta.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28e61742",
   "metadata": {},
   "source": [
    "## 2.3. vertical plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ecc2ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_bar_bm(df_prob_input.iloc[:, 0:5], \n",
    "            label_sample = ['BM_ID 1, White', 'BM_ID 2, Black', 'BM_ID 3, Asian', 'BM_ID 4, Indian', 'BM_ID 5, Asian'],\n",
    "            loc_legend = \"right\", bool_save_plot = True, \n",
    "            path_plot = \"../figure/figure3/\",\n",
    "            name_plot = \"figure3_BM_1to5\",\n",
    "            total_width = 0.8, \n",
    "            size_plot_x = 3.9, size_plot_y = 1.9, pad_x_tick = 0.035)\n",
    "\n",
    "plot_bar_bm(df_prob_input.iloc[:, [5,7,6,8]], \n",
    "         label_sample = ['ET1 (scRNA-seq)', 'ET1 (WGS)', 'ET2 (scRNA-seq)', 'ET2 (WGS)'],\n",
    "         loc_legend = \"left\", bool_save_plot = True, \n",
    "         path_plot = \"../figure/figure3/\",\n",
    "         name_plot = \"figure3_BM_6and7\", \n",
    "         total_width = 0.9, \n",
    "         size_plot_x = 3.25, pad_x_tick = 0.035)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12570f8e",
   "metadata": {},
   "source": [
    "## 2.4. horizontal plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15bd957",
   "metadata": {},
   "source": [
    "### 2.4.1. BM1 ~ 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ebe4c25",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_barhplot(df_prob_input.iloc[:, 0:5].T, df_meta.iloc[0:5, :], label_tick = 'ID_ind',\n",
    "             bool_save_plot = True,\n",
    "             path_plot = \"../figure/figure3/\",\n",
    "             name_plot = \"figure3_BM1to5\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "640c8b52",
   "metadata": {},
   "source": [
    "### 2.4.2. BM6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d02293d",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_barhplot(df_prob_input.iloc[:, [5,7]].T, df_meta.iloc[[5,7], :], label_tick = 'ID_ind',\n",
    "             bool_save_plot = True,\n",
    "             path_plot = \"../figure/figure3/\",\n",
    "             name_plot = \"figure3_BM6\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b542850",
   "metadata": {},
   "source": [
    "### 2.4.3. BM7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204600e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_barhplot(df_prob_input.iloc[:, [6,8]].T, df_meta.iloc[[6,8], :], label_tick = 'ID_ind',\n",
    "             bool_save_plot = True,\n",
    "             path_plot = \"../figure/figure3/\",\n",
    "             name_plot = \"figure3_BM7\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80c9de42",
   "metadata": {},
   "source": [
    "# 3. admixtures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88a4616d",
   "metadata": {},
   "source": [
    "## 3.1 functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c52f916b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# horizontal stacked plots\n",
    "def get_barhplot_admixed(df_prob_input, df_meta_input, label_tick = 'GT.read', \n",
    "                         bool_save_plot = False, path_plot = \"\", labelpad_y = 50,\n",
    "                 size_plot_x = 6, size_plot_y = 1.5, name_plot = \"\", min_prob_label = 0.09):\n",
    "    \n",
    "    test_bool_1 = (df_prob_input.index == df_meta_input.index).all()\n",
    "    if (not(test_bool_1)):\n",
    "        print(f'<get_barhplot> WARNING: index of df_prob_input not df_meta_input equal')\n",
    "    \n",
    "    vec_bool_row_not_NA = ~df_prob_input.isna().all(axis = 1)\n",
    "    n_row_NA = df_prob_input.shape[0] - sum(vec_bool_row_not_NA)\n",
    "    if (n_row_NA != 0):\n",
    "        print(f'<get_barhplot> WARNING: number of NA rows: {n_row_NA}')\n",
    "    \n",
    "    df_prob_input = df_prob_input[vec_bool_row_not_NA]\n",
    "    df_meta_input = df_meta_input[vec_bool_row_not_NA]\n",
    "    \n",
    "    # reverse order\n",
    "    df_meta_input = df_meta_input.iloc[::-1]\n",
    "    df_prob_input = df_prob_input.iloc[::-1]\n",
    "        \n",
    "    n_sample = df_prob_input.shape[0]\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(size_plot_x, size_plot_y))  # Create a figure and an axes.\n",
    "\n",
    "    # ... after creating your plot ...\n",
    "    adjust_left_margin(fig, ax, fixed_margin=0.2)\n",
    "\n",
    "    # Set the background color\n",
    "    fig.set_facecolor('white')  # Set the figure background to white\n",
    "    ax.set_facecolor('white')   # Set the axes background to white\n",
    "    \n",
    "    # Initial position of the left edge of the bar\n",
    "    df = df_prob_input\n",
    "    lefts = [0] * len(df)\n",
    "    \n",
    "    label_legend = list(dict_color_super.keys())\n",
    "    color_legend = list(dict_color_super.values())\n",
    "    \n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0 - 0.5, df.shape[0] - 0.5)\n",
    "    \n",
    "    scale_factor = calculate_scale_factor(fig, ax)\n",
    "    \n",
    "    print(f\"scale_factor {scale_factor}\")\n",
    "    \n",
    "    height_bar = 0.12 * scale_factor\n",
    "    vec_coord_y = np.arange(n_sample) * height_bar * 1.25\n",
    "\n",
    "    for i_sample, column in enumerate(df.columns):\n",
    "        widths = df[column]\n",
    "        temp_super = get_super(column)\n",
    "        rects = ax.barh(vec_coord_y, widths, left = lefts, height = height_bar, label=column, \n",
    "                        color = dict_color_super[temp_super], edgecolor = 'black')\n",
    "        # Update the left edge for the next segment\n",
    "        lefts = [left + value for left, value in zip(lefts, widths)]\n",
    "        \n",
    "        for rect in rects:\n",
    "            width = rect.get_width()\n",
    "            if width > min_prob_label:  # only if there's a visible segment\n",
    "                ax.text(rect.get_x() + width/2, rect.get_y() + rect.get_height()/2,\n",
    "                        column, ha = 'center', va = 'center', fontsize = size_font, \n",
    "                        color = 'white')\n",
    "                    \n",
    "    ax.set_yticks(vec_coord_y)\n",
    "    ax.set_yticklabels(list(df_meta_input[label_tick]), fontsize = size_font)\n",
    "    ax.set_xticks([])\n",
    "    \n",
    "    ax.tick_params(axis = 'y', which = 'major', pad = 0.01)\n",
    "    ax.tick_params(axis = 'x', which = 'major', pad = 0.01)\n",
    "    \n",
    "    for label in ax.get_yticklabels():\n",
    "        label.set_x(label.get_position()[0] + 0.03)\n",
    "        \n",
    "    \n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if (bool_save_plot):\n",
    "        fn_path_exists(path_plot)\n",
    "        print(f'<get_barhplot> saving at path_plot: {path_plot}')\n",
    "        plt.savefig(f'{path_plot}bone-marrow_horizontal_{name_plot}.pdf', \n",
    "                    format = 'pdf', dpi = 1200, bbox_inches = 'tight')\n",
    "        plt.savefig(f'{path_plot}bone-marrow_horizontal_{name_plot}.jpeg', \n",
    "                    format = 'jpeg', dpi = 1200, bbox_inches = 'tight')\n",
    "        \n",
    "# vertical plots\n",
    "def plot_bar_admixed(df_prob, label_sample = \"\", loc_legend = \"left\", \n",
    "                     bool_save_plot = False, path_plot = \"\", name_plot = \"\", size_plot_x = 3.0,\n",
    "                     size_plot_y = 2.0, pad_x_tick = 0.035, total_width = 1):\n",
    "    n_sample = df_prob.shape[1]\n",
    "    df_prob = df_prob.loc[vec_pop_ordered]\n",
    "    fig, ax = plt.subplots(figsize=(size_plot_x, size_plot_y))  # Create a figure and an axes.\n",
    "\n",
    "    # Set the background color\n",
    "    fig.set_facecolor('white')  # Set the figure background to white\n",
    "    ax.set_facecolor('white')   # Set the axes background to white\n",
    "    \n",
    "    # Bar plot settings\n",
    "    # total_width = 1  # Total width for a group of bars\n",
    "    bar_width = total_width / n_sample\n",
    "    \n",
    "    # bar_width = 1.0 / n_sample  # Width of individual bars\n",
    "    # total_width = n_sample * bar_width  # Total width for a group of bars\n",
    "    ind = np.arange(len(df_prob))\n",
    "        \n",
    "    # Plot each bar\n",
    "    for i, col in enumerate(df_prob.columns):\n",
    "        if (i > 0):\n",
    "            i_color = i + 2\n",
    "            if (i > 1):\n",
    "                i_color = i_color + 1\n",
    "        else:\n",
    "            i_color = i\n",
    "                \n",
    "        offset = (i + 1/2) * bar_width - total_width / 2  # Center bars around each tick\n",
    "        ax.bar(ind + offset, df_prob[col], \n",
    "               bar_width, # bar width # bar_width + n_sample/20\n",
    "               color = plt.cm.viridis(np.linspace(0, 1, n_sample + 3))[i_color], \n",
    "               edgecolor = plt.cm.viridis(np.linspace(0, 1, n_sample + 3))[i_color], \n",
    "               label = col, linewidth = 0)\n",
    "        \n",
    "    # Set x-ticks to the center of each group of bars\n",
    "    ax.set_xticks(ind)\n",
    "    ax.set_xticklabels(df_prob.index, rotation='vertical', fontsize=size_font)\n",
    "\n",
    "    plt.xlim(-0.5, 25.5)\n",
    "    plt.xlabel('Populations', fontsize = size_font)\n",
    "    plt.ylabel('Predicted Probabilities', fontsize = size_font)\n",
    "    plt.yticks(np.arange(0, 1.1, 0.2))  # Adjust the range and step as needed\n",
    "\n",
    "    plt.tick_params(axis = 'both', which = 'major', labelsize = size_font)\n",
    "\n",
    "    # get handles and labels from the old legend\n",
    "    old_handles, old_labels = ax.get_legend_handles_labels()\n",
    "\n",
    "    # create new handles for the new legend entries\n",
    "    new_handles = [patches.Rectangle((0, 0), 1, 1, color = dict_color_super[region_i]) for region_i in vec_sup]\n",
    "    new_labels = vec_sup_full\n",
    "\n",
    "    if label_sample == \"\":\n",
    "        label_sample = [\"sample\" + str(i) for i in range(1, n_sample + 1)]\n",
    "    \n",
    "    labels_donor = label_sample\n",
    "    # combine old and new handles and labels\n",
    "    handles = old_handles + new_handles\n",
    "    # labels = labels_donor + new_labels\n",
    "    labels = labels_donor[:n_sample]\n",
    "\n",
    "    # create the new legend\n",
    "    legend_bar = ax.legend(handles, labels, loc = 'upper ' + loc_legend, fontsize = size_font,\n",
    "                           frameon=True, facecolor='white', edgecolor = \"black\")\n",
    "    \n",
    "    # legend_bar.set_title(title_plot + \" (% of GT Reads)\", prop={\"size\": size_font}) \n",
    "    ax.add_artist(legend_bar)\n",
    "\n",
    "    for label in ax.get_xticklabels():  \n",
    "        pop_temp = label.get_text()\n",
    "        color_temp = dict_color_super[df_meta_unique[df_meta_unique['SUP'] == pop_temp]['POP'].values[0]]\n",
    "        label.set_bbox(dict(facecolor = color_temp, edgecolor = 'None', alpha = 0.5, pad = 0.5))  # change color as needed\n",
    "    \n",
    "    # Adjusting space between the x-axis and its tick labels\n",
    "    ## 1. Reduce pad to bring labels closer\n",
    "    ax.tick_params(axis = 'x', which = 'major', pad = 0.001)\n",
    "\n",
    "    ## 2. Manually adjust the position of x-axis tick labels\n",
    "    for label in ax.get_xticklabels():\n",
    "        label.set_y(label.get_position()[1] + pad_x_tick)\n",
    "        \n",
    "    # Reduce empty space inside the plot area\n",
    "    ax.margins(x = 0.001)  # Reduce the x margins (Adjust the value as needed)\n",
    "\n",
    "    ax.spines['top'].set_visible(True)\n",
    "    ax.spines['right'].set_visible(True)\n",
    "    ax.spines['bottom'].set_visible(True)\n",
    "    ax.spines['left'].set_visible(True)\n",
    "\n",
    "    ax.spines['top'].set_color('black')\n",
    "    ax.spines['right'].set_color('black')\n",
    "    ax.spines['bottom'].set_color('black')\n",
    "    ax.spines['left'].set_color('black')\n",
    "\n",
    "    ax.spines['top'].set_linewidth(1)\n",
    "    ax.spines['right'].set_linewidth(1)\n",
    "    ax.spines['bottom'].set_linewidth(1)\n",
    "    ax.spines['left'].set_linewidth(1)\n",
    "    \n",
    "    ax.yaxis.set_ticks([])\n",
    "    ax.yaxis.set_ticklabels([]) \n",
    "    \n",
    "    # Reduce padding and adjust layout\n",
    "    plt.tight_layout(pad=0.4, w_pad=0.3, h_pad=0.3)  # Adjust padding parameters as needed\n",
    "\n",
    "    # Adjust subplot parameters to reduce unnecessary margins\n",
    "    plt.subplots_adjust(left=0.05, right=0.975, top=0.975, bottom=0.10)\n",
    "    \n",
    "    if (bool_save_plot):\n",
    "        fn_path_exists(path_plot)\n",
    "        print(f'<plot_bar_admixed> saving at path_plot: {path_plot}')\n",
    "        plt.savefig(f'{path_plot}bone-marrow_vertical_{name_plot}.pdf', \n",
    "                    format = 'pdf', dpi = 1200, bbox_inches = 'tight')\n",
    "        plt.savefig(f'{path_plot}bone-marrow_vertical_{name_plot}.jpeg', \n",
    "                    format = 'jpeg', dpi = 1200, bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b6c2a4",
   "metadata": {},
   "source": [
    "## 3.2. plot admixtures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35472433",
   "metadata": {},
   "source": [
    "### 3.2.1. horizontal\n",
    "\n",
    "run the following code block to get the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e986462",
   "metadata": {},
   "outputs": [],
   "source": [
    "# path_admixed = '../data/results_admixed/'\n",
    "# fn_path_exists(path_admixed)\n",
    "# for name_mixed in vec_mixed:\n",
    "#     file_df_prob_read = f'{path_admixed}{name_mixed}/df_probabilities.tsv'\n",
    "#     fn_file_exists(file_df_prob_read)\n",
    "#     df_prob_read = pd.read_csv(file_df_prob_read, \n",
    "#                                sep = '\\t', index_col = 0)\n",
    "#     df_prob = df_prob_read.loc[vec_pop_ordered].T\n",
    "#     df_meta = pd.DataFrame({'origin': list(df_prob.index),\n",
    "#                             'GT.read': ['100.0'] * 3})\n",
    "#     df_meta['origin'] = df_meta['origin'].str.replace('_', ' + ')\n",
    "#     df_meta.index = list(df_prob.index)\n",
    "\n",
    "#     get_barhplot_admixed(df_prob_input = df_prob, df_meta_input = df_meta,\n",
    "#                 label_tick = 'origin', bool_save_plot = True, \n",
    "#                 path_plot = '../figure/sfigure4/',\n",
    "#                 name_plot = name_mixed, \n",
    "#                 size_plot_x = 2.4, size_plot_y = 1.2, min_prob_label = 0.09)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b07bfb0b",
   "metadata": {},
   "source": [
    "### 3.2.2. vertical\n",
    "\n",
    "run the following code block to get the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aba1e082",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for name_mixed in vec_mixed:\n",
    "#     file_df_prob_read = f'{path_admixed}{name_mixed}/df_probabilities.tsv'\n",
    "#     fn_file_exists(file_df_prob_read)\n",
    "#     df_prob_read = pd.read_csv(file_df_prob_read, sep = '\\t', index_col = 0)\n",
    "#     df_prob = df_prob_read.loc[vec_pop_ordered]\n",
    "#     list_col = list(df_prob.columns)\n",
    "#     label_sample = [col.replace('_', ' + ') for col in list_col]\n",
    "\n",
    "#     plot_bar_admixed(df_prob, \n",
    "#              label_sample = label_sample,\n",
    "#              loc_legend = \"right\", bool_save_plot = True, \n",
    "#              path_plot = '../figure/sfigure4/', name_plot = name_mixed, \n",
    "#              size_plot_x = 2.2, \n",
    "#              size_plot_y = 1.2, \n",
    "#              pad_x_tick = 0.07, total_width = 0.8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "725af8b9",
   "metadata": {},
   "source": [
    "## 3.3. make synthetic admixtures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7207ab76",
   "metadata": {},
   "source": [
    "The sample code below was used to make the synthetic admixture individuals. To run this code, the user needs to download and process the 1000 Genomes Project (1kGP) data. In this gitlab, the 1kGP data itself is not provided as it is already publicly available and because it is too large."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a54e94b",
   "metadata": {},
   "source": [
    "### 3.3.1. functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c67098",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mix_df(df1, df2, n_mut):\n",
    "    \"\"\"\n",
    "    Mixes two dataframes by randomly shuffling indices and then taking a subset from each dataframe\n",
    "    to create a new mixed dataframe.\n",
    "\n",
    "    :param df1: The first dataframe to mix.\n",
    "    :param df2: The second dataframe to mix.\n",
    "    :param n_mut: The number of mutations, which determines the range of indices to shuffle.\n",
    "    :return: A new dataframe consisting of a random mix of rows from df1 and df2.\n",
    "    \"\"\"\n",
    "    # Generate a list of numbers from 0 to n_mut-1\n",
    "    numbers = list(range(n_mut))\n",
    "\n",
    "    # Shuffle the list of numbers randomly\n",
    "    random.shuffle(numbers)\n",
    "\n",
    "    # Calculate the midpoint of the list to split it into two halves\n",
    "    midpoint = len(numbers) // 2\n",
    "    # Take the first half of the shuffled numbers for df1\n",
    "    list1 = numbers[:midpoint]\n",
    "    # Take the second half of the shuffled numbers for df2\n",
    "    list2 = numbers[midpoint:]\n",
    "\n",
    "    # Concatenate the selected rows from df1 and df2 into a new dataframe\n",
    "    # and sort it by the original index to maintain order\n",
    "    return(pd.concat([df1.iloc[list1], df2.iloc[list2]]).sort_index())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33d96764",
   "metadata": {},
   "source": [
    "### 3.3.2. data creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15ac0a0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(52014)\n",
    "\n",
    "list_sub_intra1 = ['ACB', 'ASW', 'CLM', 'PEL', 'CDX', 'CHS', 'CEU', 'FIN', 'BEB', 'GIH']\n",
    "list_sub_intra2 = ['YRI', 'MSL', 'PUR', 'MXL', 'KHV', 'JPT', 'TSI', 'IBS', 'STU', 'PJL']\n",
    "\n",
    "list_sub_inter1 = ['ESN', 'GWD', 'LWK', 'MSL', 'CLM', 'PEL', 'PUR', 'JPT', 'CHB', 'GBR']\n",
    "list_sub_inter2 = ['MXL', 'CDX', 'CEU', 'BEB', 'CHB', 'IBS', 'GIH', 'TSI', 'PJL', 'STU']\n",
    "\n",
    "dict_vec_GT_mixed_intra = {}\n",
    "dict_df_prob_intra = {}\n",
    "dict_ld_intra = {}\n",
    "dict_p_site_present_intra = {}\n",
    "\n",
    "# takes four minutes\n",
    "for sub1, sub2 in zip(list_sub_intra1, list_sub_intra2):\n",
    "    dict_vec_GT_mixed_intra[f'{sub1}_{sub2}'] = mix_df(dict_vec_GT_raw[f'{sub1}'],\n",
    "                                                       dict_vec_GT_raw[f'{sub2}'])\n",
    "dict_vec_GT_mixed_inter = {}\n",
    "dict_df_prob_inter = {}\n",
    "dict_ld_inter = {}\n",
    "dict_p_site_present_inter = {}\n",
    "for sub1, sub2 in zip(list_sub_inter1, list_sub_inter2):\n",
    "    dict_vec_GT_mixed_inter[f'{sub1}_{sub2}'] = mix_df(dict_vec_GT_raw[f'{sub1}'],\n",
    "                                                       dict_vec_GT_raw[f'{sub2}'])\n",
    "    \n",
    "dict_vec_GT_mixed_intra = {}\n",
    "dict_df_prob_intra = {}\n",
    "dict_ld_intra = {}\n",
    "dict_p_site_present_intra = {}\n",
    "\n",
    "# takes four minutes\n",
    "for sub1, sub2 in zip(list_sub_intra1, list_sub_intra2):\n",
    "    dict_vec_GT_mixed_intra[f'{sub1}_{sub2}'] = mix_df(dict_vec_GT_raw[f'{sub1}'], \n",
    "                                                 dict_vec_GT_raw[f'{sub2}'])    \n",
    "dict_vec_GT_mixed_inter = {}\n",
    "dict_df_prob_inter = {}\n",
    "dict_ld_inter = {}\n",
    "dict_p_site_present_inter = {}\n",
    "for sub1, sub2 in zip(list_sub_inter1, list_sub_inter2):\n",
    "    dict_vec_GT_mixed_inter[f'{sub1}_{sub2}'] = mix_df(dict_vec_GT_raw[f'{sub1}'], dict_vec_GT_raw[f'{sub2}'])\n",
    "df_prob_intra = pd.concat(dict_df_prob_intra).reset_index(level = 1, drop = True)\n",
    "df_prob_intra = df_prob_intra.T.loc[vec_pop_ordered].T\n",
    "\n",
    "df_meta_intra = pd.DataFrame.from_dict(dict_p_site_present_intra, orient = 'index', columns=['GT.read'])\n",
    "df_meta_intra['origin'] = [f'{sub1} + {sub2}' for sub1, sub2 in zip(list_sub_intra1, list_sub_intra2)]\n",
    "\n",
    "df_prob_inter = pd.concat(dict_df_prob_inter).reset_index(level = 1, drop = True)\n",
    "df_prob_inter = df_prob_inter.T.loc[vec_pop_ordered].T\n",
    "\n",
    "df_meta_inter = pd.DataFrame.from_dict(dict_p_site_present_inter, orient='index', columns=['GT.read'])\n",
    "df_meta_inter['origin'] = [f'{sub1} + {sub2}' for sub1, sub2 in zip(list_sub_inter1, list_sub_inter2)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
